File: algorithms/medium/longest_subarray_of_1s_after_deleting_one_element/solution.py
class Solution:
    def longestSubarray(self, nums: list) -> int:
        zeros = []
        for i in range(0, len(nums)):
            if nums[i] == 0:
                zeros.append(i)

        if len(zeros) == 0:
            return len(nums)-1

        max = 0
        for z in range(0, len(zeros)):
            start, end = 0, len(nums)-1
            if z != 0:
                start = zeros[z-1]+1
            if z != len(zeros)-1:
                end = zeros[z+1]-1

            if end-start > max:
                max=end-start

        return max

cases = [
    [1,1,0,1],
    [0,1,1,1,0,1,1,0,1],
    [1,1,1]
]

answers = [3,5,2]

solution = Solution()
for idx, v in enumerate(cases):
    answer = solution.longestSubarray(v)
    if answer != answers[idx]:
        print(f'n={v} Wrong!')
        print(f'output:{v}, answer:{answers}')
        break
    
    print(f'n={v} Correct!')

File: algorithms/medium/longest_subarray_of_1s_after_deleting_one_element/solution.go
package main

import "fmt"

func longestSubarray(nums []int) int {
	zeros := make([]int, 0)
	for i := 0; i < len(nums); i++ {
		if nums[i] == 0 {
			zeros = append(zeros, i)
		}
	}

	if len(zeros) == 0 {
		return len(nums) - 1
	}

	max := 0
	for z := 0; z < len(zeros); z++ {
		start, end := 0, len(nums)-1
		if z != 0 {
			start = zeros[z-1] + 1
		}
		if z != len(zeros)-1 {
			end = zeros[z+1] - 1
		}

		if end-start > max {
			max = end - start
		}
	}

	return max
}

func main() {
	fmt.Println(longestSubarray([]int{1, 1, 0, 1}) == 3)
	fmt.Println(longestSubarray([]int{0, 1, 1, 1, 0, 1, 1, 0, 1}) == 5)
	fmt.Println(longestSubarray([]int{1, 1, 1}) == 2)
}
